core/src/test/java/org/infinispan/functional/AbstractFunctionalOpTest.java

```diff
@@ -283,9 +283,9 @@ public static VisitableCommand getCurrent() {
       }
 
       @Override
-      protected BasicInvocationStage handleDefault(InvocationContext ctx, VisitableCommand command) throws Throwable {
+      protected InvocationStage handleDefault(InvocationContext ctx, VisitableCommand command) throws Throwable {
          current.set(command);
-         return invokeNext(ctx, command).handle((rCtx, rCommand, rv, t) -> current.remove());
+         return invokeNext(ctx, command).whenComplete(ctx, command, (rCtx, rCommand, rv, t) -> current.remove());
```

Comment:
@rvansa I think Pedro meant that the callback is going to be called on a different thread, so it's going to remove the wrong thread-local. We could use `try { invokeNext(ctx, command).get(); } finally { current.remove() }` to make the removal more reliable, but that still won't guarantee that the lambda that reads the thread-local is invoked on the same thread.